migrations/admin_events.go

```diff
@@ -0,0 +1,151 @@
+package migrations
+
+import (
+	"fmt"
+
+	"github.com/mongodb/anser"
+	"github.com/mongodb/anser/db"
+	"github.com/mongodb/anser/model"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	eventCollection       = "event_log"
+	adminDataType         = "ADMIN"
+	eventTypeKey          = "e_type"
+	eventTypeBanner       = "BANNER_CHANGED"
+	eventTypeTheme        = "THEME_CHANGED"
+	eventTypeServiceFlags = "SERVICE_FLAGS_CHANGED"
+	eventTypeValueChanged = "CONFIG_VALUE_CHANGED"
+)
+
+type EventDataOld struct {
+	ResourceType string       `bson:"r_type"`
+	User         string       `bson:"user"`
+	OldVal       string       `bson:"old_val"`
+	NewVal       string       `bson:"new_val"`
+	OldFlags     ServiceFlags `bson:"old_flags"`
+	NewFlags     ServiceFlags `bson:"new_flags"`
+}
+
+type ServiceFlags struct {
+	TaskDispatchDisabled         bool `bson:"task_dispatch_disabled"`
+	HostinitDisabled             bool `bson:"hostinit_disabled"`
+	MonitorDisabled              bool `bson:"monitor_disabled"`
+	NotificationsDisabled        bool `bson:"notifications_disabled"`
+	AlertsDisabled               bool `bson:"alerts_disabled"`
+	TaskrunnerDisabled           bool `bson:"taskrunner_disabled"`
+	RepotrackerDisabled          bool `bson:"repotracker_disabled"`
+	SchedulerDisabled            bool `bson:"scheduler_disabled"`
+	GithubPRTestingDisabled      bool `bson:"github_pr_testing_disabled"`
+	RepotrackerPushEventDisabled bool `bson:"repotracker_push_event_disabled"`
+	CLIUpdatesDisabled           bool `bson:"cli_updates_disabled"`
+	GithubStatusAPIDisabled      bool `bson:"github_status_api_disabled"`
+}
+
+type EventDataNew struct {
+	ResourceType string           `bson:"r_type"`
+	User         string           `bson:"user"`
+	Section      string           `bson:"section"`
+	Changes      ConfigDataChange `bson:"changes"`
+}
+
+type ConfigDataChange struct {
+	Before interface{} `bson:"before"`
+	After  interface{} `bson:"after"`
+}
+
+type Settings struct {
+	Banner      string `bson:"banner"`
+	BannerTheme string `bson:"banner_theme"`
+}
+
+func adminEventRestructureGenerator(env anser.Environment, db string, limit int) (anser.Generator, error) {
+	const migrationName = "admin_event_restructure"
+
+	if err := env.RegisterManualMigrationOperation(migrationName, makeAdminEventMigration(db)); err != nil {
+		return nil, err
+	}
+
+	opts := model.GeneratorOptions{
+		NS: model.Namespace{
+			DB:         db,
+			Collection: eventCollection,
+		},
+		Limit: limit,
+		Query: bson.M{
+			"r_id":        "",
```

Comment:
To clarify, you're trying to get all events with where r_id exists, and it's an empty string right, plus the r_type thing?

Also, it it possible to restrict this query further with timestamps? `ts` is indexed too

Reply:
Yep, that's how the admin events are stored currently. We don't want to restrict the query anymore, and it's fast enough with the current indices